ilastik/workflows/carving/opPreprocessing.py

```diff
@@ -182,40 +182,36 @@ def setupOutputs(self):
 
     def execute(self, slot, subindex, roi, result):
         assert roi.stop - roi.start == self.Output.meta.shape, "Blockwise Watershed must be run on the entire volume."
-        input_image = self.Input(roi.start, roi.stop).wait()
-        volume_feat = input_image[0, ..., 0]
-        result_view = result[0, ..., 0]
-        # handle the special case of the Request threadpool not having any workers
-        max_workers = max(1, Request.global_thread_pool.num_workers)
-        with Timer() as watershedTimer:
-            # 3d watersheds
-            if self.Input.meta.getTaggedShape()["z"] > 1:
-                logger.info("Run block-wise watershed in 3d")
-                if self.DoAgglo.value:
-                    result_view[...], max_id = watershed_and_agglomerate(
-                        volume_feat,
-                        max_workers=max_workers,
+
+        if self.Input.meta.getTaggedShape()["z"] > 1:
+            idx = numpy.s_[0, ..., 0]
+        else:
+            idx = numpy.s_[0, ..., 0, 0]
+
+        # Don't adjust roi to idx because idx should only drop singleton dims.
+        input_ = self.Input(roi.start, roi.stop).wait()[idx]
+        assert input_.ndim in 2, 3
+
+        with Timer() as timer:
+            logger.info("Run block-wise watershed in %sd", input_.ndim)
+
+            if self.DoAgglo.value:
+                result[idx], max_id = watershed_and_agglomerate(
+                        input_,
+                        max_workers=max(1, Request.global_thread_pool.num_workers),
                         size_regularizer=self.SizeRegularizer.value,
                         reduce_to=self.ReduceTo.value,
                     )
-                else:
-                    result_view[...], max_id = vigra.analysis.watershedsNew(volume_feat)
-            # 2d watersheds
+            elif input_.ndim == 3:
+                result[idx], max_id = vigra.analysis.watersheds(input_)
+            elif input_.ndim == 2:
+                result[idx], max_id = vigra.analysis.watershedsNew(input_)
             else:
-                logger.info("Run block-wise watershed in 2d")
-                if self.DoAgglo.value:
-                    result_view[...], max_id = watershed_and_agglomerate(
-                        volume_feat[:, :, 0],
-                        max_workers=max_workers,
-                        size_regularizer=self.SizeRegularizer.value,
-                        reduce_to=self.ReduceTo.value,
-                    )
-                else:
-                    result_view[...], max_id = vigra.analysis.watersheds(volume_feat[:, :, 0])
+                raise AssertionError("Invalid ndim")
```

Comment:
I don't like the use of `AssertionError` in general. Just think about catching/handling it. `AssertionError` is normally generated by `assert` statements. You'd have to `except AssertionError` to do any handling there which is never a good idea: those might get optimized away when python is invoked with the appropriate flags. In any case the error message seems to imply, that you want to raise a `ValueError`. But that also doesn't really fit the if statement. I would subclass exception here.

Reply:
The assertion error is there just to make the if-else chain complete, ndim should be checked before anyway. Also, this will go away because we only need one, new watershed implementation in both cases.